src/theory/datatypes/sygus_simple_sym.cpp

```diff
@@ -0,0 +1,597 @@
+ *   d_children[1]:
+ *     d_req_type = A
+ * This trie is satisfied by sygus types that have a constructor whose builtin
+ * kind is PLUS and whose argument types are both A.
+ */
+class ReqTrie
+{
+ public:
+  ReqTrie() : d_req_kind(UNDEFINED_KIND) {}
+  /** the children of this node */
+  std::map<unsigned, ReqTrie> d_children;
+  /** the (builtin) kind required by this node */
+  Kind d_req_kind;
+  /** the type that this node is required to be */
+  TypeNode d_req_type;
+  /** the constant required by this node */
+  Node d_req_const;
+  /** print this trie */
+  void print(const char* c, int indent = 0)
+  {
+    if (d_req_kind != UNDEFINED_KIND)
+    {
+      Trace(c) << d_req_kind << " ";
+    }
+    else if (!d_req_type.isNull())
+    {
+      Trace(c) << d_req_type;
+    }
+    else if (!d_req_const.isNull())
+    {
+      Trace(c) << d_req_const;
+    }
+    else
+    {
+      Trace(c) << "_";
+    }
+    Trace(c) << std::endl;
+    for (std::map<unsigned, ReqTrie>::iterator it = d_children.begin();
+         it != d_children.end();
+         ++it)
+    {
+      for (int i = 0; i <= indent; i++)
+      {
+        Trace(c) << "  ";
+      }
+      Trace(c) << it->first << " : ";
+      it->second.print(c, indent + 1);
+    }
+  }
+  /**
+   * Are the requirements of this trie satisfied by sygus type tn?
+   * tdb is a reference to the sygus term database.
+   */
+  bool satisfiedBy(quantifiers::TermDbSygus* tdb, TypeNode tn)
+  {
+    if (!d_req_const.isNull())
+    {
+      if (!tdb->hasConst(tn, d_req_const))
+      {
+        return false;
+      }
+    }
+    if (!d_req_type.isNull())
+    {
+      Trace("sygus-sb-debug")
+          << "- check if " << tn << " is type " << d_req_type << std::endl;
+      if (tn != d_req_type)
+      {
+        return false;
+      }
+    }
+    if (d_req_kind != UNDEFINED_KIND)
+    {
+      Trace("sygus-sb-debug")
+          << "- check if " << tn << " has " << d_req_kind << std::endl;
+      std::vector<TypeNode> argts;
+      if (tdb->canConstructKind(tn, d_req_kind, argts))
+      {
+        bool ret = true;
+        for (std::map<unsigned, ReqTrie>::iterator it = d_children.begin();
+             it != d_children.end();
+             ++it)
+        {
+          if (it->first < argts.size())
+          {
+            TypeNode tnc = argts[it->first];
+            if (!it->second.satisfiedBy(tdb, tnc))
+            {
+              return false;
+            }
+          }
+          else
+          {
+            return false;
+          }
+        }
+        if (!ret)
+        {
+          return false;
+        }
+      }
+      else
+      {
+        return false;
+      }
+    }
+    return true;
+  }
+  /** is this the empty (trivially satisfied) trie? */
+  bool empty()
+  {
+    return d_req_kind == UNDEFINED_KIND && d_req_const.isNull()
+           && d_req_type.isNull() && d_children.empty();
+  }
+};
+
+// this function gets all easy redundant cases, before consulting rewriters
```

Comment:
move this comment to the documentation in the header

Reply:
its old, ill remove.